test_conformance/events/test_event_dependencies.cpp

```diff
@@ -447,6 +447,14 @@ int test_event_enqueue_wait_for_events_run_test(
 
     test_error(error, "clEnqueueReadBuffer failed");
 
+    error = clFinish(queues[0]);
```

Comment:
The `clEnqueueReadBuffer` call prior to this to `queues[0]` is blocking, I'm not sure if this `clFinish` is required for correct behaviour of this user code according to the OpenCL spec

Reply:
Yes, but the read buffer command is only dependent on `last_loop_event`, while the other ND-range kernel and the two write buffer commands have no successors.

Reply:
ah makes sense, thanks

Reply:
Sorry, I'm still not following why the `clFinish` calls are required.

Is this needed to ensure the correct data is read by `clEnqueueReadBuffer`?  (I don't think so, because the calls to `clFinish` occur after the call to `clEnqueueReadBuffer`.)

I suppose some commands may not have executed before the command-queue (or queues) is released, but that should be fine, too, and I don't think there is any requirement that a command-queue is empty before it is released.

Can you please say a little more what the calls to `clFinish` are doing?

Reply:
You are right that there is no such requirement, but if there is no sync point for all commands, they might not complete until function `main` returns, and deleting them when the program exits might cause undefined finalization scenarios.

Reply:
I've opened https://github.com/KhronosGroup/OpenCL-Docs/issues/1238 to discuss the spec issue.

Reply:
Thanks!

I'm fine merging these changes as-is while the spec issue is under discussion.  Adding in the calls to `clFinish` definitely isn't wrong, though (depending how the spec issue gets resolved) they may be unnecessary.

Reply:
Same, fine with these merging as-is while spec issue is under discussion.